### PR TITLE
[New Model]: Shared Aspect for Quantity with Unit of Measure

### LINKED ISSUE #303: [New Model]: Shared Aspect for common units of measure
## Model Description
The WeekBasedMaterialDemand model contains a unit of measure definition, that we want to reuse at least in PlannedProductionOutput (PR #295 ). Therefore I propose a shared aspect.

- weight: GRM (gram), KGM (kilogram), TNE (metric ton), STN (ton / short ton), ONZ (ounce), LBR (pound)
- length: CMT(centimetre), MTR (metre), KTM (kilometre), INH (inch), FOT (foot), YRD (yard)
- area: CMK (square centimetre), MTK (square metre), INK (square inch), FTK (square foot), YDK (square yard)
- volume: CMQ (cubic centimetre), MTQ (cubic metre), INQ (cubic inch), FTQ (cubic foot), YDQ (cubic yard), MLT (millilitre), LTR (litre), HLT (hectolitre)
- misc: H87 (pieces), SET (set of things), PR (pair), ZP (pages), KWH (kilowatt hour)

## Relevant Standards
[UN recommendation 20](https://tfig.unece.org/contents/recommendation-20.htm) (same as SAMM UnitReference).

## Example Data
KGM
H87

## MS1 Criteria

- [x] The proposed aspect model does not exist already
- [x] The proposed aspect model does not extend an existing aspect but introduces completely new functionality
- [x] The proposal references a Catena-X use case
- [x] Relevant standards are mentioned/linked
- [x] A first draft of the data structure is provided in an example

### PR DESCRIPTION
**Pull request by tom-rm-meyer-ISST**
## Description
As we reused it in PR #295 and it originates from week based material demand, we propose to outsource it and reuse it. We're planning to use it in ProductStock, too. 
I created the MS2 first. MS1 will be added afterwards, so that we can directly do the MS3 next week, if others confirm.

Renaming might be a case.

Closes  #303 

## MS2 Criteria
(to be filled out by PR reviewer)
- [x] the model **validates** with the BAMM SDS SDK in the version specified in the Readme.md of this repository by the time of the MS2 check  (e.g., 'java -jar bamm-cli.jar -i \<path-to-aspect-model\> -v ). The  SAMM CLI is available [here](https://eclipse-esmf.github.io/esmf-developer-guide/tooling-guide/samm-cli.html) and in [GitHub](https://github.com/eclipse-esmf/esmf-sdk/releases/tag/v2.3.1)
- [x] use **Camel-Case** (e.g., "MyModelElement" or "TimeDifferenceGmtId", when in doubt follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)
- [x] the identifiers for all model elements **start with a capital letter** except for properties
- [x] the identifier for **properties starts with a small letter**
- [x] all model elements **at least contain the fields "preferred name" and "description"** in English language. The description must be comprehensible. It is not required to write full sentences but style should be consistent over the whole model
- [x] Property and the referenced Characteristic should not have the same name
- [x] the versioning in the URN **follows semantic versioning**, where minor version bumps are backwards compatible and major version bumps are not backwards compatible. 
- [x] use **abbreviations only when necessary** and if these are sufficiently common
- [x] **avoid redundant prefixes in property names** (consider adding properties to an enclosing Entity or even adapt the namespace of the model elements, e.g., instead of having two properties `DismantlerId` and `DismantlerName` use an Entity `Dismantler` with the properties `name` and `id` or use a URN like `io.catenax.dismantler:0.0.1`)
- [x] fields `preferredName` and `description` are not the same
- [x] **`preferredName` should be human readable** and follow normal orthography (e.g., no camel case but normal word separation)
- [x] name of aspect is singular except if it only has one property which is a Collection, List or Set. In theses cases, the aspect name is plural.
- [x] units are referenced from the BAMM unit catalog whenever possible
- [x] **use constraints** to make known constraints from the use case explicit in the aspect model 
- [x] when relying on **external standards**, they are referenced through a **"see"** element
- [x] all properties with an [simple type](https://eclipse-esmf.github.io/samm-specification/2.0.0/datatypes.html) have an example value
- [x] metadata.json exists with status "release"
- [x] generated json schema validates against example json payload
- [x] file RELEASE_NOTES.md exists and contains entries for proposed model changes 
- [x] all contributors to this model are mentioned in copyright header of model file

## MS3 Criteria
(to be filled out by semantic modeling team before merge to main-branch)
- [x] All required reviewers have approved this PR (see reviewers section)
- [x] The new aspect (version) will be implemented by at least one data provider
- [x] The new aspect (version) will be consumed by at least one data consumer
- [x] There exists valid test data
- [x] In case of a new (incompatible) major version to an existing version, a migration strategy has been developed
- [x] The model has at least version '1.0.0'
- [x] The release date in the Release Note is set to the date of the MS3 approval

### CODE REVIEW

**Comment by tom-rm-meyer-ISST** [comment]
@bs-jokri, in PR #295 we had a unit of measure defintion that was reused from DCM. Further usage is likely in our UC. Could you please do the review or delegate?

File: io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#######################################################################
+
+@prefix samm: <urn:samm:org.eclipse.esmf.samm:meta-model:2.0.0#>.
+@prefix samm-c: <urn:samm:org.eclipse.esmf.samm:characteristic:2.0.0#>.
+@prefix samm-e: <urn:samm:org.eclipse.esmf.samm:entity:2.0.0#>.
+@prefix unit: <urn:samm:org.eclipse.esmf.samm:unit:2.0.0#>.
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
+@prefix : <urn:samm:io.catenax.unit_of_measure:1.0.1#>.
```
Comment (catroest) [inline io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl:34]: prefix doen't fit folder structure

Comment (tom-rm-meyer-ISST) [reply io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl]: Thanks! updated.

File: io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "Unit of measure commonly used in demand and supply communication of items."@en;
+    samm:characteristic :UnitOfMeasureCharacteristic;
+    samm:exampleValue "H87".
+:UnitOfMeasureCharacteristic a samm-c:Enumeration;
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "The Enumeration lists a selection of allowed units of measure referenced by their official code according to Codes for Units of Measure Used in International Trade."@en;
+    samm:dataType xsd:string;
+    samm:see <https://tfig.unece.org/contents/recommendation-20.htm>;
+    samm-c:values ("GRM" "KGM" "TNE" "STN" "ONZ" "LBR" "CMT" "MTR" "KTM" "INH" "FOT" "YRD" "CMK" "MTK" "INK" "FTK" "YDK" "CMQ" "MTQ" "INQ" "FTQ" "YDQ" "MLT" "LTR" "HLT" "H87" "SET" "PR" "ZP" "KWH").
```
Comment (catroest) [inline io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl:52]: There are over 3500 meantioned in the excel. Is this enumeration complete? Should the aspect be more strictly formulated if there is a specific focus?

**Comment by catroest** [comment]
Is it possible to split up the shared aspect into seperate characterisitcs for each "category" of units (misc, volume, e.g...). Therefore please also look at the Digital Product Passport Aspect. There is the part for "physical dimensions" where we would use your aspect. We categeorized it into linear units, capacity units and mass units. Maybe you can also name (some of) your categories like this.

Does it make sense to split up?

**Comment by tom-rm-meyer-ISST** [comment]
> Is it possible to split up the shared aspect into seperate characterisitcs for each "category" of units (misc, volume, e.g...). Therefore please also look at the Digital Product Passport Aspect. There is the part for "physical dimensions" where we would use your aspect. We categeorized it into linear units, capacity units and mass units. Maybe you can also name (some of) your categories like this. EDIT: or use for the naming of the "categories" from each dimension: https://confluence.catena-x.net/display/ARTIV/ART4DCM-892+Unit+of+Measure+Handling
> 
> Does it even make sense to split up?

We heavily investigeted how to split it up further and reuse it. From one semantic point of view it would make sense to group common units of measure (e.g. mass, area, etc.). 

The problem is that I want to use the unit of measure for a quanitty with a more flexible but still constrained set of units (enumeration). I only see the possibility to achieve this by using an unitReference and constraint the curie value respectively using a regex. I pushed a second property to visualize the change and disucss this.

I would have loved to define specific enums and combine them into one but didn't find a way.

What do you think? I guess this approach two would definitely need a renaming.

**Comment by catroest** [comment]
@tom-rm-meyer-ISST I would propose to make a meeting regarding this aspect. I would also point out, that I would like to not only have the unit, but also the affilitated value. Otherwise for the value you would always need to make a secon property in you own model even if you use the shared aspect.

**Comment by tom-rm-meyer-ISST** [comment]
I like the idea! I'll have a meeting tomorrow with the DCM colleagues. I also sent out the invite to you. We could later discuss, if the appointment is not possible.

**Comment by tom-rm-meyer-ISST** [comment]
Updated to a shared quantity. As aligned with other partners, we created different unit types and one unit combining those.

@catroest I'm not sure if the properties we reused from DigitalProductPassport should be renamed to be a Quantity same as the countable and item ones.

Please do a review of the update. 

Regarding "unitOfMeasure2": We'll incorporate or remove the approach according to feedback of @bs-jokri

File: io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+:unitOfMeasure a samm:Property;
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "Unit of measure commonly used in demand and supply communication of items."@en;
+    samm:characteristic :UnitOfMeasureCharacteristic;
+    samm:exampleValue "H87".
+:UnitOfMeasureCharacteristic a samm-c:Enumeration;
+    samm:preferredName "Unit of Measure"@en;
+    samm:description "The Enumeration lists a selection of allowed units of measure referenced by their official code according to Codes for Units of Measure Used in International Trade."@en;
+    samm:dataType xsd:string;
+    samm:see <https://tfig.unece.org/contents/recommendation-20.htm>;
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply io.catenax.shared.unit_of_measure/1.0.0/UnitOfMeasure.ttl]: updated to curie as this is possible now

**Comment by tom-rm-meyer-ISST** [comment]
Hey together (at least @catroest , @nhaenis)

Today with @bs-jokri  I clarified that:
- the catch all item quantity is qiute OK, as there is no enumeration superset functionality in SAMM. We need to keep an eye on it during update.
-  we can use curie as datatype for enum values.

As requested in issue #319 I added cycles (countable) and seconds (time). I also added time values from seconds to days (@nhaenis feel free to give feedback).

We're still in clarification on how to proceed with the empty value. I'll let you know soon.

**Comment by tom-rm-meyer-ISST** [comment]
> ### Validation Report for io.catenax.shared.quantity/1.0.0/Quantity.ttl
> #### Validation failed:

locally it validates in aspect model editor 4.5.2

File: io.catenax.shared.quantity/1.0.0/Quantity.ttl [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+@prefix samm-e: <urn:samm:org.eclipse.esmf.samm:entity:2.0.0#>.
+@prefix unit: <urn:samm:org.eclipse.esmf.samm:unit:2.0.0#>.
+@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
+@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
+@prefix : <urn:samm:io.catenax.shared.quantity:1.0.0#>.
+
+:Quantity a samm:Aspect;
+    samm:preferredName "Quantity with Unit of Measure"@en;
+    samm:description "The aspect may be reused whenever any kind of Quantity and / or unit of measure is used. One can use all defined units via the item unit or differenciate via mass, miscellaneous, countable, linear, area and volume units."@en;
```
Comment (agg3fe) [inline io.catenax.shared.quantity/1.0.0/Quantity.ttl:39]: typo in `differenciate`

**Comment by agg3fe** [comment]
@tom-rm-meyer-ISST  please correct the small typo. After that we are good for MS2.

**Comment by tom-rm-meyer-ISST** [comment]
> @tom-rm-meyer-ISST please correct the small typo. After that we are good for MS2.

@agg3fe fixed + updated RELEASE_NOTES for MS3 on monday.

**Comment by tom-rm-meyer-ISST** [comment]
Raised SDK issue [#259](https://github.com/eclipse-esmf/esmf-semantic-aspect-meta-model/issues/259) due to github action failure.

**Comment by chris-volk** [comment]
> Raised SDK issue [#259](https://github.com/eclipse-esmf/esmf-semantic-aspect-meta-model/issues/259) due to github action failure.

Thanks. I moved it to the SDK repository (copy content, mutual link & close original) as you can in the links

**Review by agg3fe — APPROVED** [review]
LGTM

**Comment by agg3fe** [comment]
MS3 given on the condition that @catroest also approves the changes.

**Comment by tom-rm-meyer-ISST** [comment]
@catroest is currently not available. Should we wait till she is?

**Comment by agg3fe** [comment]
As discussed in the weekly meeting, this model Must be merged by Friday, as this is blocking other use cases.

**Pull request merged**